core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,953 @@
+		if (!idx && hashdata->block_data_filled) {
+			/* Use hashdata->block_data if present to */
+			memcpy(data_block, hashdata->block_data,
+			       hashdata->block_data_filled);
+			/* clen = remaining size to fill a block */
+			clen = DCP_SHA_BLOCK_SIZE - hashdata->block_data_filled;
+			/* Fill the buffer to its maximum */
+			memcpy(data_block + hashdata->block_data_filled, data,
+			       clen);
+			/* Actual size in the local buffer (<= DCP_SHA_BLOCK) */
+			size = clen + hashdata->block_data_filled;
+			/* reset block_data_filled for last iteration */
+			hashdata->block_data_filled = 0;
+		} else if (idx != nb_block - 1) {
+			/* Fill data block completely */
+			memcpy(data_block, data + clen, DCP_SHA_BLOCK_SIZE);
+			clen += DCP_SHA_BLOCK_SIZE;
+			size = DCP_SHA_BLOCK_SIZE;
+		} else if (idx == nb_block - 1) {
+			/*
+			 * Save data in hashdata structure to have data to
+			 * submit during final operation
+			 */
+			memcpy(hashdata->block_data, data + clen, len - clen);
+			hashdata->block_data_filled = len - clen;
+			ret = TEE_SUCCESS;
+			break;
+		}
+
+		desc->buff_size = size;
+		ret = dcp_run(&hashdata->dcp_data);
+		/*
+		 * After the first call of DCP operation we need to unset
+		 * DCP_CONTROL0_HASH_INIT
+		 */
+		if (desc->ctrl0 & DCP_CONTROL0_HASH_INIT)
+			desc->ctrl0 &= ~DCP_CONTROL0_HASH_INIT;
+	}
+
+	dcp_unlock_channel(hashdata->dcp_data.channel);
+
+	return ret;
+}
+
+TEE_Result dcp_sha_do_final(struct dcp_hashdata *hashdata, uint8_t *digest,
+			    size_t digest_size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	size_t payload_size = 0;
+	unsigned int i = 0;
+	struct dcp_descriptor *desc = NULL;
+	struct dcp_align_buf payload = { };
+
+	if (!hashdata) {
+		EMSG("Bad parameters, hashdata is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	if (!hashdata->op_state.init) {
+		EMSG("Error, please call dcp_sha_do_init() before");
+		return TEE_ERROR_ACCESS_DENIED;
+	}
+	if (!&hashdata->dcp_data.desc) {
+		EMSG("Bad parameters, hashdata->dcp_data.desc is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	if (!digest) {
+		EMSG("Bad parameters, digest is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	if (digest_size < hash_alg[hashdata->hash_id].size) {
+		EMSG("Digest buffer size is to small, should be %" PRId32,
+		     hash_alg[hashdata->hash_id].size);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	desc = &hashdata->dcp_data.desc;
+	payload_size = hash_alg[hashdata->hash_id].size;
+	ret = dcp_calloc_align_buf(&payload, payload_size);
+	if (ret != TEE_SUCCESS) {
+		EMSG("payload allocation failed");
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	/* Set work packet for last iteration */
+	desc->ctrl0 |= DCP_CONTROL0_HASH_TERM;
+	desc->src_buffer = hashdata->block_data;
+	desc->buff_size = hashdata->block_data_filled;
+	desc->payload = payload.data;
+	if (hashdata->hash_check == DCP_SHA_CHECK) {
+		desc->ctrl0 |= DCP_CONTROL0_HASH_CHECK;
+		/*
+		 * Fill payload buffer with check value, for some reason DCP
```

Comment:
"for some reason" doesn't sound very reassuring.
Is this the same thing as "contrary to the hardware documentation, the result is flipped"?

Reply:
I will check the reference manual on my side to check if this behavior is expected or not.